acceptance/tests/pxp-module-puppet/restart_pxp_agent_during_non_blocking_run.rb

```diff
@@ -0,0 +1,154 @@
+    default:   { exec { 'sleep': command => '/bin/sleep #{SECONDS_TO_SLEEP} || /bin/true', } }
+  }
+}
+MODULEPP
+    on(master, "chmod 644 #{module_manifest}")
+
+    step 'Link the environment\'s temp dir to the actual Puppet environmentpath'
+    on(master, "rm -rf #{master_environment_path}/#{ENVIRONMENT_NAME}") # Ensure folder does not pre-exist, or ln will create link inside it
+    on(master, "ln -s #{tmp_environment_dir} #{master_environment_path}/#{ENVIRONMENT_NAME}")
+  end
+
+  step 'Ensure each agent host has pxp-agent running and associated' do
+    agents.each do |agent|
+      on agent, puppet('resource service pxp-agent ensure=stopped')
+      create_remote_file(agent, pxp_agent_config_file(agent), pxp_config_json_using_puppet_certs(master, agent).to_s)
+      on agent, puppet('resource service pxp-agent ensure=running enable=true')
+      show_pcp_logs_on_failure do
+        assert(is_associated?(master, "pcp://#{agent}/agent"),
+               "Agent #{agent} with PCP identity pcp://#{agent}/agent should be associated with pcp-broker")
+      end
+    end
+  end
+
+  applicable_agents.each do |agent|
+    agent_identity = "pcp://#{agent}/agent"
+    transaction_id = nil
+    step "Make a non-blocking puppet run request on #{agent}" do
+      provisional_responses = rpc_non_blocking_request(master, [agent_identity],
+                                      'pxp-module-puppet', 'run',
+                                      {:env => [], :flags => ['--onetime',
+                                                              '--no-daemonize',
+                                                              "--environment", "'#{ENVIRONMENT_NAME}'"]})
+      assert_equal(provisional_responses[agent_identity][:envelope][:message_type],
+                   "http://puppetlabs.com/rpc_provisional_response",
+                   "Did not receive expected rpc_provisional_response in reply to non-blocking request")
+      transaction_id = provisional_responses[agent_identity][:data]["transaction_id"]
+    end
+
+    step 'Wait to ensure that Puppet has time to execute manifest' do
+      begin
+        retry_on(agent, "ps -ef | grep 'bin/sleep' | grep -v 'grep'", {:max_retries => 15, :retry_interval => 2})
+      rescue
+        fail("After triggering a puppet run on #{agent} the expected sleep process did not appear in process list")
+      end
+    end
+
+    step "Restart pxp-agent service on #{agent}" do
+      on agent, puppet('resource service pxp-agent ensure=stopped')
+      on agent, puppet('resource service pxp-agent ensure=running')
+    end
+
+    step 'Signal sleep process to end so Puppet run will complete' do
+      agent['platform'] =~ /osx/ ?
+        command = "ps -e -o pid,comm | grep sleep | sed 's/^[^0-9]*//g' | cut -d\\  -f1" :
+        command = "ps -ef | grep 'bin/sleep' | grep -v 'grep' | grep -v 'true' | sed 's/^[^0-9]*//g' | cut -d\\  -f1"
```

Comment:
I'm not a fan of using the tertiary operator in void context for side-effects like this.  I think it'd read clearer as:

```
command = agent['platform'] =~ /osx/ ? "OSX cli" : "elsewhere cli"
```